crates/aptos-speculative/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+/// event and manifests the desired behavior as a side effect (returning nothing). This occurs
+/// if and when a speculative event actually gets finalized.
+pub trait SpeculativeEvent {
+    fn dispatch(&self);
+}
+
+// A type alias for event storage container.
+type EventStore<T> = Vec<CachePadded<Mutex<LazyClearVec<T>>>>;
+
+/// A struct that stores speculative events for transactions, indexed by an usize. Allows
+/// clearing the speculative events for a specific transaction or all transactions, and flushing
+/// the (non-cleared) events when the executions are finalized. The underlying storage must be
+/// sized fit the indices of transactions, set by the new(num_txns) call for initialization.
+pub struct SpeculativeEvents<E: Send> {
+    events: OnceCell<EventStore<E>>,
+}
+
+impl<E: Send + SpeculativeEvent + 'static> SpeculativeEvents<E> {
+    // Returns a ref to the current storage of all events if its length is sufficiently large.
+    fn events_with_checked_length(&self, min_length: usize) -> anyhow::Result<&EventStore<E>> {
+        let events = self.events.get().unwrap();
+        if events.len() < min_length {
+            anyhow::bail!(
+                "speculative events storage len = {} < required {} (was not sized appropriately)",
+                events.len(),
+                min_length
+            );
+        }
+        Ok(events)
+    }
+
+    /// Create a new storage for recording speculative events by transactions indexed 0..num_txns.
+    pub fn new(num_txns: usize) -> Self {
+        Self {
+            events: OnceCell::with_value(
+                (0..num_txns)
+                    .map(|_| CachePadded::new(Mutex::new(LazyClearVec::default())))
+                    .collect(),
+            ),
+        }
+    }
+
+    /// Error means that the underlying speculative event storage was not properly sized to store
+    /// an event for the provided transaction index.
+    pub fn record(&self, txn_idx: usize, event: E) -> anyhow::Result<()> {
+        let events = self.events_with_checked_length(txn_idx + 1)?;
+
+        events[txn_idx].lock().push(event);
+        Ok(())
+    }
+
+    /// Logically clears events recorded so far for a given transaction.
+    pub fn clear_txn_events(&self, txn_idx: usize) -> anyhow::Result<()> {
+        let events = self.events_with_checked_length(txn_idx + 1)?;
+        events[txn_idx].lock().clear();
+        Ok(())
+    }
+
+    /// Logically clears all events.
+    pub fn clear_all_events(&self) {
+        // TODO: Parallelize if needed.
+        let events = self.events.get().unwrap();
+        for event in events {
+            event.lock().clear();
+        }
+    }
+
+    /// Flush the stored events asynchronously by spawning global rayon threads.
+    pub fn flush(mut self) {
+        let events = self.events.take().unwrap();
+
+        rayon::spawn(move || {
+            events.par_iter().with_min_len(25).for_each(|m| {
```

Comment:
made it const. it might still be hardcoded in places across the codebase, because it's a totally optional thing, e.g. we can just omit with_min_length and it would use 1, it's just trying to batch a bit better when the individual iterator lambdas may be too lightweight.